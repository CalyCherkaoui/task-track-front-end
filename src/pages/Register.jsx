/* eslint-disable jsx-a11y/label-has-associated-control */
import React, { useState, useRef } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import Form from 'react-validation/build/form';
import Input from 'react-validation/build/input';
import CheckButton from 'react-validation/build/button';
import { isEmail } from 'validator';

import { register } from '../actions/authentication';

// eslint-disable-next-line consistent-return
const required = (value) => {
  if (!value) {
    return (
      <div className="alert alert-danger" role="alert">
        Required to Signin!
      </div>
    );
  }
};

// eslint-disable-next-line consistent-return
const validEmail = (value) => {
  if (!isEmail(value)) {
    return (
      <div className="alert alert-danger" role="alert">
        Oops! It is not a valid email! Try agin!
      </div>
    );
  }
};

// eslint-disable-next-line consistent-return
const validUsername = (value) => {
  if (value.length < 3 || value.length > 20) {
    return (
      <div className="alert alert-danger" role="alert">
        Oops! Your username must be more than 3 characters. Please try again!
      </div>
    );
  }
};

// eslint-disable-next-line consistent-return
const validPassword = (value) => {
  if (value.length < 6 || value.length > 40) {
    return (
      <div className="alert alert-danger" role="alert">
        The password must be between 6 and 40 characters.
      </div>
    );
  }
};

const Register = () => {
  const form = useRef();
  const checkBtn = useRef();

  const [username, setUsername] = useState('');
  const [email, setEmail] = useState('');
  const [password, setPassword] = useState('');
  const [successful, setSuccessful] = useState(false);

  const { message } = useSelector((state) => state.message);
  const dispatch = useDispatch();

  const onChangeUsername = (e) => {
    const username = e.target.value;
    setUsername(username);
  };

  const onChangeEmail = (e) => {
    const email = e.target.value;
    setEmail(email);
  };

  const onChangePassword = (e) => {
    const password = e.target.value;
    setPassword(password);
  };

  const handleRegister = (e) => {
    e.preventDefault();
    setSuccessful(false);
    form.current.validateAll();
    // eslint-disable-next-line no-underscore-dangle
    if (checkBtn.current.context._errors.length === 0) {
      dispatch(register(username, email, password))
        .then(() => {
          setSuccessful(true);
        })
        .catch(() => {
          setSuccessful(false);
        });
    }
  };

  return (
    <div className="col-md-12">
      <div className="card card-container">
        <img
          src="//ssl.gstatic.com/accounts/ui/avatar_2x.png"
          alt="profile-img"
          className="profile-img-card"
        />

        <Form onSubmit={handleRegister} ref={form}>
          {!successful && (
            <div>
              <div className="form-group">
                <label htmlFor="username">
                  Username:
                  <Input
                    type="text"
                    className="form-control"
                    name="username"
                    placeholder="Enter your name"
                    value={username}
                    onChange={onChangeUsername}
                    validations={[required, validUsername]}
                  />
                </label>
              </div>

              <div className="form-group">
                <label htmlFor="email">
                  Email:
                  <Input
                    type="text"
                    className="form-control"
                    name="email"
                    value={email}
                    onChange={onChangeEmail}
                    placeholder="Enter a valid Email"
                    validations={[required, validEmail]}
                  />
                </label>
              </div>

              <div className="form-group">
                <label htmlFor="password">
                  Password:
                  <Input
                    type="password"
                    className="form-control"
                    name="password"
                    value={password}
                    placeholder="Enter a strong password"
                    onChange={onChangePassword}
                    validations={[required, validPassword]}
                  />
                </label>
              </div>

              <div className="form-group">
                <button
                  className="btn btn-primary btn-block"
                  type="submit"
                >
                  Sign Up
                </button>
              </div>
            </div>
          )}

          {message && (
            <div className="form-group">
              <div className={successful ? 'alert alert-success' : 'alert alert-danger'} role="alert">
                {message}
              </div>
            </div>
          )}
          <CheckButton style={{ display: 'none' }} ref={checkBtn} />
        </Form>
      </div>
    </div>
  );
};

export default Register;
