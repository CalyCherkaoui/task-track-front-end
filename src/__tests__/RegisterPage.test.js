import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';
import { BrowserRouter as Router } from 'react-router-dom';
import { Provider } from 'react-redux';
import store from '../store';
import Register from '../pages/Register';

describe(
  'Renders correctly Register page component',
  () => {
    let container;
    beforeEach(() => {
      container = shallow(
        <Provider store={store}>
          <Router>
            <Register />
          </Router>
        </Provider>,
      );
    });

    it(
      'Contains \'Sign me up!\' button',
      () => {
        expect(container.find('Sign me up')).toBeTruthy();
      },
    );

    it(
      'Matches Register snapshot',
      () => {
        const rendered = renderer.create(
          <Provider store={store}>
            <Router>
              <Register />
            </Router>
          </Provider>,
        );

        expect(rendered).toMatchSnapshot();
      },
    );
  },
);
