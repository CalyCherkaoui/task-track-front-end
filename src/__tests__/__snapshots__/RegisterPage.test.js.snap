// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders correctly Register page component Matches Register snapshot 1`] = `
<div
  className="form_wrapper box_flex_col_centered p-4"
>
  <div
    className="g-0 shadowed_small container"
  >
    <div
      className="form_header_wrapper g-0 row"
    >
      <div
        className="box_flex_col_centered g-0 col"
      >
        <p
          className="form_header_text"
        >
          Sign-up
        </p>
      </div>
    </div>
    <div
      className="form_form_wrapper g-0 row"
    >
      <div
        className="box_flex_col_centered py-4 col"
      >
        <form
          className="form_form px-3"
          onSubmit={[Function]}
        >
          <label
            className="form_label box_flex_col_centered"
            htmlFor="signup_username_input"
          >
            <input
              className="form_input"
              id="signup_username_input"
              name="username"
              onChange={[Function]}
              placeholder="Enter your username"
              required={true}
              type="text"
            />
          </label>
          <label
            className="form_label box_flex_col_centered"
            htmlFor="signup_email_input"
          >
            <input
              className="form_input"
              id="signup_email_input"
              name="email"
              onChange={[Function]}
              placeholder="Enter your email!"
              required={true}
              type="email"
            />
          </label>
          <label
            className="form_label box_flex_col_centered"
            htmlFor="signup_password_input"
          >
            <input
              className="form_input"
              id="signup_password_input"
              name="password"
              onChange={[Function]}
              placeholder="Enter your password!"
              required={true}
              type="password"
            />
          </label>
          <button
            className="form_submit"
            type="submit"
          >
            Sign me up!
          </button>
        </form>
      </div>
    </div>
  </div>
</div>
`;
